Optimization: skip tests in some circumstances

edsantiago opened the pull request:
A common pattern is to submit PRs that update only tests or docs.

When the only changes are to test/e2e, there is no point in running
test/system, and vice versa.

When the only changes are to docs, there is no point in running either.

Signed-off-by: Ed Santiago <santiago@redhat.com>

rhatdan commented:
Can you do the same for just docs?  People often forget to add [CI:DOCS] when doing a docs only change. It would be a lot nicer if the system would figure this out.

edsantiago commented:
/me thinks @rhatdan didn't read the code :wink:

rhatdan commented:
Red but did not comprehend.  Mentally did an AND instead of an OR on the skip.

cevich reviewed (commented):
I think this should be okay.  Though I'm somewhat bothered we can't make the entire task "go away" as opposed remaining in the clutter-pile.  There is a mechanism for doing something similar directly in `.cirrus.yml` with a `changesInclude()` function ([docs ref](https://cirrus-ci.org/guide/writing-tasks/#supported-functions)) that could be used with `only_if` to achieve that.  However, the `only_if` logic is currently a giant putrid hairball, so this may be the best alternative.

The only other thing that gave me a slight pause is the use of `$CIRRUS_CHANGE_IN_REPO`.  I'm not 100% positive that has a useful value in every context where `_only_changes_are_in()`.  For example on a branch of tag triggered run.  Though given the way you structured this, that would just make the tests run if there's a failure from the `git merge-base ...`.  So even this may be okay.

edsantiago commented:
Thanks, @cevich. I too wish we could do this within `.cirrus.yml` but see no sane way to do it. Here we have an imperfect solution which still (sigh) spins up VMs but at least wastes much less developer and CPU time.

Re: `$CIRRUS_CHANGE_IN_REPO` - excellent point, thank you, I've fixed that logic.

cevich commented:
Update, so here's a run on a tag: https://cirrus-ci.com/build/5093673821011968

I see in the F35 build task, it is showing a `$CIRRUS_CHANGE_IN_REPO` which matches the tagged commit but `$DEST_BRANCH` is set wrong :confused: 

@edsantiago it's critical that we only do high-reliability checks on new tags, since there's no opportunity to fix the code (w/o force-pushing).  Would you mind adding a check early on to see if `$CIRRUS_TAG` is non-empty.  If so, probably the best thing is to not check any other conditions (i.e. git merge base will probably fail), to avoid problems that can't be fixed.  Does that make sense?

cevich commented on contrib/cirrus/runner.sh:
```diff
@@ -345,6 +390,32 @@ dotest() {
+    local head base diffs
+
+    # Cirrus sets this for PRs but not cron. In cron, we never want to skip.
+    if [[ -z "$CIRRUS_CHANGE_IN_REPO" ]]; then
```
Ya this is good.  Probably should stick in a `|| [[ -n "$CIRRUS_TAG" ]]` here too.

cevich commented:
> but see no sane way to do it.

Ya, it's just too complicated, and not growing any simpler.  There is a scripting language I've been thinking of that would make these things MUCH easier.  It can partly or fully take the place of `.cirrus.yml`, and supports actual/sane conditionals and looping.  [Docs ref.](https://cirrus-ci.org/guide/programming-tasks/) if your curious. 'Course would be a big migration and would break your cirrus task-map utility :cry: But it would bring some sanity to handling things like `CI:DOCS` and other conditions (like change-based skips) without so much brain-hurt.  Anyway, something I've been thinking of.

edsantiago commented:
Validation: to test the logic, I snuck in a check-and-skip for `contrib` (like in this PR). [Cirrus results](https://cirrus-ci.com/build/5281713126375424) confirm that pretty much everything got skipped as intended, you can verify by looking at all the `00:`, `01:`, and `02:` in the left-hand column. I clicked on a few of them and looked at `main`, confirmed that they truly say `Skipping`. I did not verify each individual one.

I've removed that `contrib` check (because I think we _should_ run full CI on changes therein), and have incorporated @cevich's excellent suggestions, and have removed the Draft status. I'm happy enough with this, and would appreciate final review.

(@rhatdan , @cevich, do note that I've added many more skip checks since the last time you may have looked).

edsantiago commented on contrib/cirrus/runner.sh:
```diff
@@ -55,31 +60,66 @@ function _run_unit() {
 }
 
 function _run_apiv2() {
+    if _only_changes_are_in docs hack test/{buildah-bud,e2e,system,upgrade}
```
A much smarter person than I would come up with a way to use positive logic instead of negative, e.g. "if this is a doc- or hack- or test-only change, except for changes within test/apiv2 itself". Suggestions welcome.

cevich replied:
This is probably best left up to a human to decide, e.g. by adding `[CI:DOCS]` magic or using similar mechanism.  In extreme cases, the tasks can be manually-triggered.

cevich commented on contrib/cirrus/runner.sh:
```diff
@@ -55,31 +60,66 @@ function _run_unit() {
+    then
+        return
+    fi
+
```
There's a sneaky corner-case gotcha here: The `localintegration` target depends on both `test-binaries` and `ginkgo` which both reference `$(BUILDTAGS)`.  Maybe you see where I'm going with this...`$(BUILDTAGS)` is defined by calling a whole bunch of `hack/` scripts :cry:

Maybe this is okay?  I think the most likely consequence is someone breaks something like `hack/btrfs_tag.sh` and it merges.  Literally everybody will immediately notice as the PR universe crumbles.

I'm not sure there's a sane way to deal with this, other than removing `hack` here (and maybe elsewhere).

I'm choosing to stick my head in the sand regarding problems keeping Makefile changes in-sync with `_only_changes_are_in docs` conditionals.  I think we can rely on the Makefile being so damn scarry nobody will touch it :grin:

cevich commented on contrib/cirrus/runner.sh:
```diff
@@ -217,6 +257,11 @@ function _run_build() {
+    then
+        return
+    fi
+
```
Skipping `hack` here also makes me nervous.  Right below is a call to `hack/make-and-check-size` and I know for sure the Windows build depends on both both `hack` and `docs` :disappointed:

cevich reviewed (changes requested):
okay, I looked carefully and found a few things worth considering. My focus was on tests depending on items which are also skip conditions.  Hopefully my logic-ing is correct, it's late, I need to go home :grinning:

cevich commented:
Side-bar:  @edsantiago maybe after some soak-time, you'd consider moving `_only_changes_are_in()` into containers/automation common?  There's an opportunity to add some unit-tests there, but more importantly this function will be available for use in nearly all the other containers-org CI as well.

edsantiago commented:
@cevich, good point about `hack`. I'll remove it -- no need to be overly aggressive from the start.

Re: `automation common`, my ~~thinking~~ waffling was that I would look into using `exit` from the helper itself, so we could do:
```diff
- if _only_changes_are_in this that & the other
- then
-     return
- fi
+ _bail_if_....
```
...which is so much cleaner and more readable. It just requires a lot of work for me to confirm that `exit` is 100% safe in this context. I'll do so this morning.

edsantiago commented:
Major overhaul from yesterday: I rewrote it all in a much cleaner way. Verification: with the (temporary) addition of `contrib` to force-skip all skips, CI completed [in twenty minutes](https://cirrus-ci.com/build/6621563993718784) with lots of `Skip` messages. I've removed the `contrib`, and am happy with this, and it's ready for review.

edsantiago commented:
Without the `contrib` kludge (i.e., as ready for review), CI finished in [ninety minutes](https://cirrus-ci.com/build/5644873616326656) with no skips that I can see.

cevich commented:
> ...which is so much cleaner and more readable. It just requires a lot of work for me to confirm that exit is 100% safe in this context. I'll do so this morning.

Ya I considered suggesting this, but was also nervous.  It's probably okay, since `runner.sh` is the final call for testing operations from `.cirrus.yml`.  Nothing ever follows it or depends on anything other than it exiting zero/non-zero to indicate status.

cevich commented on contrib/cirrus/runner.sh:
```diff
@@ -345,6 +364,51 @@ dotest() {
+
+    # Filter these out from the diffs, see what's left.
+    for subdir in docs test; do
+        # || true needed because we're running with set -e
```
good catch

cevich commented on contrib/cirrus/runner.sh:
```diff
@@ -345,6 +364,51 @@ dotest() {
+        fi
+    done
+
+    # Filter these out from the diffs, see what's left.
```
This comment (WRT `for subdir in docs test; do`) is confusing to me (it specifies "what" not "why").  Shouldn't it be filtering out based on the arguments (`$@`) not hard-coded values?

edsantiago replied:
Arguments have already been filtered out (in the code immediately above). What we do now is: if this is a docs-or-test-only PR, then it's safe to skip. The `docs` and `test` hardcoding has to happen here, not in the caller: specifying them in the caller was my approach last night, and it was a nightmare.

cevich replied:
So if you remember, for a future PR or update, please fix the comment to mention something like "hard-coding docs and test to avoid overburdening caller".  Or maybe in the main function comment.

cevich commented on contrib/cirrus/runner.sh:
```diff
@@ -345,6 +364,51 @@ dotest() {
+    # Cirrus sets these for PRs but not cron. In cron, we never want to skip.
+    for v in CIRRUS_CHANGE_IN_REPO DEST_BRANCH; do
+        if [[ -z "${!v}" ]]; then
+            echo "[ _cannot do selective skip: \$$v is undefined ]"
```
Non-blocking: Would prefer use of `msg()` and/or `warn()` functions for consistency (warn() prints the LOC it was called from).

cevich reviewed (approved):
This is much better.  I'm having a hard time with some of the comments and message output - many refer explicitly to 'tests' and 'docs' but the function also accepts arguments.  Normally I expect a function to only hard-code or only use input arguments.  The reasoning behind doing both wasn't clear to me.  But this is a non-blocker because sometimes my coffee stops working earlier than I would like :grin: 

Thanks for taking this work on @edsantiago it's a much needed and useful improvement.  I'm not that concerned about the wasted 1-minute of time to spin up a VM & potential queuing for such.  Given the pre-existing complexity of `.cirrus.yml` we can leave that optimizations up to the future.

/lgtm
/hold

edsantiago commented:
New push addresses your concerns: using `msg`, and updating the comments for clarity.

cevich reviewed (approved):
Fantastic, much clearer now thanks again @edsantiago

rhatdan commented on contrib/cirrus/runner.sh:
```diff
@@ -55,31 +57,45 @@ function _run_unit() {
 }
 
 function _run_apiv2() {
+    _bail_if_test_can_be_skipped test/apiv2
```
Does this say only test APIv2 if there were changes in the test/apiv2 directory?

cevich replied:
No, it's the opposite.  Ed fixed the comment that handles the arguments:

```
# If PR touches any files in an argument directory, we cannot skip
```

cevich replied:
(and the code does what the comment says :grin:)

rhatdan replied:
But I dont think we should be skipping apiv2 tests if the test/apiv2 directory was not changed.  If the pkg/api directory or anything in pkg was changed, it could break the apiv2 tests.

cevich replied:
right, I follow you now, yes this was confusing for me too.  So the subsequent block removes all 'test/*' and 'docs/*' changes, then only blocks testing if there are no other changes.  So, if in your example some podman code was changed, the apiv2 tests **would** run.  It would only be skipped if apiv2 changes *only* included other changes under `test/*` and/or `docs/*`.  It's confusing, but I think Ed's got it right.

cevich replied:
```
    # If we still have diffs, they indicate files outside of docs & test.
    # It is not safe to skip.
    if [[ -n "$diffs" ]]; then
        return 0
    fi
```

edsantiago replied:
@rhatdan let's see if this clarifies: `_bail_if_test_can_be_skipped TEST/FOO` will bail if TEST/FOO can be skipped. A test can be skipped if:
1. the PR does not modify any files under TEST/FOO; and
2. the PR does not modify any files other than in `docs` and `test`.

That is: if PR touches only files in `test/e2e`, we can skip `test/system`.

If PR touches any file outside of `docs` and `test`, any file at all whatsoever, the full CI suite will run. Does that help?

rhatdan replied:
Got it now I understand.

rhatdan commented:
/lgtm

cevich commented:
/hold cancel
